Implement basic command bars

klondikedragon opened the pull request:
This is a first implementation of command bars as described in #231.

A `CommandBar` control provides quick access to common tasks. This could be application-level or page-level commands.

Refer to [the spec](https://docs.microsoft.com/en-us/windows/apps/design/controls/command-bar).

The current implementation does not implement the open/closed state, nor does it implement the secondary commands "more" button with dynamic overflow of items into the secondary commands menu.

Also, during implementation of the example, I realized that I needed control over the `primary` and `controller` properties of the `ListView` inside of the `TreeView`, so this also adds these properties to `TreeView`.

Here is what the examples look like with the various overflow behaviors:

![image](https://user-images.githubusercontent.com/88254524/159105153-0f7d5f7a-b432-40a0-a0b9-62b4213c8b56.png)

And here is what it looks like alongside the large `TreeView`:

![image](https://user-images.githubusercontent.com/88254524/159105166-0af75829-d5bf-4875-9902-a4736fd8cb58.png)

## Pre-launch Checklist

- [X] I have updated `CHANGELOG.md` with my changes <!-- REQUIRED --> 
- [X] I have run "optimize/organize imports" on all changed files
- [X] I have added/updated relevant documentation

henry2man commented:
I am very interested in this component. I also have an implementation, different but perhaps it can contribute things.For example, in my version I distinguish two groups of buttons, one of which is automatically hidden when there is no horizontal space.If you want we can mark it as a draft and spend a few days reviewing it together.

bdlukaa commented on lib/src/controls/surfaces/commandbar.dart:
```diff
@@ -0,0 +1,178 @@
+/// Buttons are the most common control to put within a [CommandBar].
+/// They are composed of an (optional) icon and an optional
+class CommandBarButton extends StatelessWidget {
+  final IconData? icon;
```
Should be a `Widget`

bdlukaa commented on lib/src/controls/surfaces/commandbar.dart:
```diff
@@ -0,0 +1,178 @@
+
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      margin: margin,
```
```suggestion
    return Padding(
      padding: margin,
```

bdlukaa commented on lib/src/controls/surfaces/commandbar.dart:
```diff
@@ -0,0 +1,178 @@
+
+/// The type of wrapping used for the items inside of the CommandBar.
+///
+///   * [CommandBarOverflowBehavior.scrolling] will cause items to scroll horizontally.
+///   * [CommandBarOverflowBehavior.noWrap] will expand the size of the CommandBar based
+///      on the size of the contained items.
+///   * [CommandBarOverflowBehavior.wrap] will wrap items onto additional lines as needed.
+///   * [CommandBarOverflowBehavior.clip] will keep items on one line and clip as needed.
```
This can be added to the items themselves

bdlukaa commented on lib/src/controls/surfaces/commandbar.dart:
```diff
@@ -0,0 +1,178 @@
+      required this.children,
+      this.wrapType = CommandBarOverflowBehavior.scrolling,
+      this.parentVerticalScrollController})
+      : _isExpanded = !(wrapType == CommandBarOverflowBehavior.noWrap),
```
```suggestion
      : _isExpanded = wrapType != CommandBarOverflowBehavior.noWrap,
```

bdlukaa commented on lib/src/controls/surfaces/commandbar.dart:
```diff
@@ -0,0 +1,178 @@
+      {Key? key,
+      required this.children,
+      this.wrapType = CommandBarOverflowBehavior.scrolling,
+      this.parentVerticalScrollController})
```
trailing comma

bdlukaa commented on lib/src/controls/surfaces/commandbar.dart:
```diff
@@ -0,0 +1,178 @@
+      this.onLongPress,
+      this.focusNode,
+      this.autofocus = false,
+      this.label})
```
trailing comma

bdlukaa commented on CHANGELOG.md:
```diff
@@ -9,6 +9,8 @@ Date format: DD/MM/YYYY
 - TreeView now implemented with ListView and optionally allows vertical scrolling ([#255](https://github.com/bdlukaa/fluent_ui/pull/225))
 - TreeViewItem now has custom primary key (`value` field)
 - Added `onSelectionChanged` callback to `TreeView`
+- Implement `CommandBar` (dynamic overflow not yet supported)
```
```suggestion
- Implement basic `CommandBar` ([#232](https://github.com/bdlukaa/fluent_ui/pull/232))
```

bdlukaa commented on lib/src/utils/horizontal_scroll_view.dart:
```diff
@@ -0,0 +1,80 @@
+  Widget build(BuildContext context) {
+    return Listener(
+      onPointerSignal: widget.mouseWheelScrolls
+          ? (event) {
+              if (event is PointerScrollEvent) {
+                _controller.animateTo(_controller.offset + event.scrollDelta.dy,
+                    duration: const Duration(milliseconds: 100),
+                    curve: Curves.ease);
+                if (widget.parentVerticalScrollController != null &&
+                    widget
+                        .parentVerticalScrollController!.positions.isNotEmpty) {
+                  widget.parentVerticalScrollController!.jumpTo(widget
+                          .parentVerticalScrollController!
+                          .positions
+                          .last
+                          .pixels -
+                      event.scrollDelta.dy);
+                }
+              }
+            }
```
What does this do?

klondikedragon replied:
It was an ugly hack that I am not proud of! I've figured out the proper way to do this using `PointerSignalResolver` and was able to delete all references to `parentVerticalScrollController`. It now captures the mouse wheel scrolling and prevents propagation of it without having to specify any surrounding scroll controllers.

bdlukaa reviewed (changes requested):
Added a basic review. Documentation and formatting needs to be improved

klondikedragon commented:
@bdlukaa - addressed comments and added `CommandBarSeparator` (corollary to `PaneItemSeparator`).

@henry2man - all help would be welcome! this could be added as an enhancement to this (whether in this PR or a follow-on) by:
* Adding a new type of `CommandBarOverflowBehavior` (something like `dynamicOverflow`)
* Adding a new widget `CommandBar` widget field `secondaryItems` (defaults to empty)
* In build if `secondaryItems` is not empty to automatically add a button using the `FluentIcons.more` icon that when clicked will present a flyout with the `secondaryItems`
* Adding logic to somehow calculate during build or layout of `CommandBar` (if overflow mode is `dynamicOverflow`) which of the normal items should instead be considered part of `secondaryItems`

Do you have any code to reference that demonstrates how to figure out which of the items can fit vs which ones should be hidden into a secondary menu?

henry2man commented:
> Do you have any code to reference that demonstrates how to figure out which of the items can fit vs which ones should be hidden into a secondary menu?

I have a very simple code: a width parameter in px that is evaluated during the execution of the "build" method. A  [LayoutBuilder](https://api.flutter.dev/flutter/widgets/LayoutBuilder-class.html) could be used: 

"if width is less than x then hide buttons and show menú button"

klondikedragon commented:
@henry2man @bdlukaa -- After some research, I figured out how to implement a fully featured command bar, including:
* Support for separate primary vs secondary command bar items, with secondary items shown in a flyout menu behind a "more" button
* Dynamic overflow behavior to put overflowing primary command bar items into the secondary area (flyout menu)
* A "compact" mode for primary command bar items, which is automatically selected if the width of the command bar is less than a specified breakpoint (off by default).
* Support for right-aligned command bars (the fluent UI spec seems to prefer this in some contexts).

This was implemented with a new `DynamicOverflow` layout widget that supports both horizontal and vertical layouts. The code for this was carefully adapted from the standard `Wrap` widget. This widget could potentially be useful in the future to implement a "more" button for horizontal or vertical navbar pane items (instead of scrolling).

In the example app, with additional examples for command bars added, it was getting cluttered on the others pane, so I added another navbar pane. I could move it back to others if you want, it was just a little awkward testing the wrapping behavior with a lot of stuff above it.

All the new functionality is best illustrated with a video:
![fluent_ui_commandbar_dynamic_overflow](https://user-images.githubusercontent.com/88254524/159151744-cde0d8e5-687d-478e-854d-67b896496397.gif)

bdlukaa commented:
SO COOL! This will open possibilities for the same thing on NavigationView! I'll review ASAP

bdlukaa commented on lib/src/controls/surfaces/commandbar.dart:
```diff
@@ -0,0 +1,552 @@
+    this.overflowItemBuilder,
+    this.overflowBehavior = CommandBarOverflowBehavior.dynamicOverflow,
+    this.compactBreakpointWidth,
+    this.mainAxisAlignment = MainAxisAlignment.start,
+    this.crossAxisAlignment = CrossAxisAlignment.center,
```
on #181, Right to Left Support was added. The current `Directionality` should be taken into account

klondikedragon replied:
Tight-to-left Directionality is already considered in the widgets used inside the `CommandBar` (whether it's `Row` or `DynamicOverflow`). I just needed to change the text "right-aligned" to "end-aligned" in the example so it's not confusing when right-to-left mode is enabled:
![image](https://user-images.githubusercontent.com/88254524/159275955-05eee4da-61ce-495e-82ee-a779f5b3f220.png)

For the secondary flyout menu content, I did notice that `ListTile` needed its padding fixed in RTL mode, so I fixed that.

bdlukaa commented on lib/src/controls/surfaces/commandbar.dart:
```diff
@@ -0,0 +1,552 @@
+            "CommandBar does not support CrossAxisAlignment.stretch");
+      case CrossAxisAlignment.baseline:
+        throw UnsupportedError(
+            "CommandBar does not support CrossAxisAlignment.baseline");
```
```suggestion
            'CommandBar does not support CrossAxisAlignment.baseline');
```

bdlukaa commented on lib/src/controls/surfaces/commandbar.dart:
```diff
@@ -0,0 +1,552 @@
+  /// This applies the widgetBuilder callback to the internally built
+  /// widget before returning.
+  Widget build(BuildContext context, CommandBarItemDisplayMode displayMode) {
+    var w = buildWidget(context, displayMode);
+    if (widgetBuilder != null) {
+      w = widgetBuilder!(context, displayMode, w);
+    }
```
Currently, `buildWidget` will always be called. This could cause performance issues and memory leak. With the suggestion, `buildWidget` will only be called if necessary
```suggestion
    if (widgetBuilder != null) {
      return widgetBuilder!(context, displayMode, w);
    } else {
      return buildWidget(context, displayMode);
    }
```

klondikedragon replied:
Now that this logic has been moved into a separate type of item, `CommandBarBuilderItem`, this logic will only be used when this type of item is explicitly used, which requires a non-null callback, so it should solve for this as well.

bdlukaa commented on lib/src/controls/surfaces/commandbar.dart:
```diff
@@ -0,0 +1,552 @@
+abstract class CommandBarItem with Diagnosticable {
+  final Key? key;
+
+  /// Specify to allow customization of the built widget for a given
+  /// display mode. For example, to wrap a widget in a [Tooltip].
+  final CommandBarItemWidgetBuilder? widgetBuilder;
```
I don't like this. The dev can extend a `CommandBarItem` and do it on the `buildWidget` method.

Something like this can be seen on https://github.com/bdlukaa/fluent_ui/blob/master/example/lib/main.dart#L300-L339

bdlukaa reviewed (changes requested):
Awesome job!

bdlukaa reviewed (approved):
Awesome work! LGTM

WinXaito commented:
Incredible !

bdlukaa commented:
@klondikedragon Hello! I'm having issues with https://github.com/bdlukaa/fluent_ui/pull/277. It uses `DynamicOverflow`, but I need to ensure the current selected item is being displayed. Can you help me on that? If you want, we can talk on discord :)
